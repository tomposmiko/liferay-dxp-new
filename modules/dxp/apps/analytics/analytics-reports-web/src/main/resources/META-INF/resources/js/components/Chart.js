/**
 * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
 *
 * The contents of this file are subject to the terms of the Liferay Enterprise
 * Subscription License ("License"). You may not use this file except in
 * compliance with the License. You can obtain a copy of the License by
 * contacting Liferay, Inc. See the License for the specific language governing
 * permissions and limitations under the License, including but not limited to
 * distribution rights of the Software.
 */

import ClayLoadingIndicator from '@clayui/loading-indicator';
import className from 'classnames';
import {useIsMounted} from 'frontend-js-react-web';
import PropTypes from 'prop-types';
import React, {useContext, useEffect, useMemo} from 'react';
import {
	CartesianGrid,
	Legend,
	Line,
	LineChart,
	ReferenceDot,
	Tooltip,
	XAxis,
	YAxis,
} from 'recharts';

import {
	useAddDataSetItems,
	useChangeTimeSpanKey,
	useChartState,
	useDateTitle,
	useIsPreviousPeriodButtonDisabled,
	useNextTimeSpan,
	usePreviousTimeSpan,
	useSetLoading,
} from '../context/ChartStateContext';
import ConnectionContext from '../context/ConnectionContext';
import {StoreContext, useHistoricalWarning} from '../context/StoreContext';
import {generateDateFormatters as dateFormat} from '../utils/dateFormat';
import {numberFormat} from '../utils/numberFormat';
import {ActiveDot as CustomActiveDot, Dot as CustomDot} from './CustomDots';
import CustomTooltip from './CustomTooltip';
import TimeSpanSelector from './TimeSpanSelector';

const CHART_COLORS = {
	analyticsReportsHistoricalReads: '#50D2A0',
	analyticsReportsHistoricalViews: '#4B9BFF',
	cartesianGrid: '#E7E7ED',
	publishDate: '#2E5AAC',
};

const CHART_SIZES = {
	dotRadius: 4,
	fill: 'white',
	height: 220,
	lineWidth: 2,
	width: 280,
	yAxisWidth: 40,
};

const DAY_IN_MILLISECONDS = 24 * 60 * 60 * 1000;
const HOUR_IN_MILLISECONDS = 60 * 60 * 1000;

const LAST_24_HOURS = 'last-24-hours';

const METRICS_STATIC_VALUES = {
	analyticsReportsHistoricalReads: {
		color: CHART_COLORS.analyticsReportsHistoricalReads,
		iconType: 'square',
		langKey: Liferay.Language.get('reads-metric'),
	},
	analyticsReportsHistoricalViews: {
		color: CHART_COLORS.analyticsReportsHistoricalViews,
		iconType: 'circle',
		langKey: Liferay.Language.get('views-metric'),
	},
};

function keyToTranslatedLabelValue(key) {
	return METRICS_STATIC_VALUES[key]?.langKey ?? key;
}

function keyToHexColor(key) {
	return METRICS_STATIC_VALUES[key]?.color ?? '#666666';
}

function keyToIconType(key) {
	return METRICS_STATIC_VALUES[key]?.iconType ?? 'line';
}

/*
 * If a number is bigger than 1000 it will transform it to kilos
 *
 * 4 => 4
 * 4000 => '4K'
 */
function thousandsToKilosFormater(value) {
	if (value > 1000) {
		return value / 1000 + 'K';
	}

	return value;
}

function legendFormatterGenerator(
	totals,
	languageTag,
	publishedToday,
	validAnalyticsConnection
) {
	return (value) => {
		const preformattedNumber = totals[value];

		return (
			<span>
				<span
					className={`custom-${keyToIconType(value)} mr-2`}
					style={{
						backgroundColor: keyToHexColor(value),
					}}
				></span>
				<span className="mr-2 text-secondary">
					{keyToTranslatedLabelValue(value)}
				</span>
				<span className="font-weight-bold">
					{validAnalyticsConnection &&
					preformattedNumber !== null &&
					!publishedToday
						? numberFormat(languageTag, preformattedNumber)
						: '-'}
				</span>
			</span>
		);
	};
}

export default function Chart({
	dataProviders = [],
	languageTag,
	publishDate,
	timeSpanOptions,
}) {
	const {validAnalyticsConnection} = useContext(ConnectionContext);

	const [{publishedToday}] = useContext(StoreContext);

	const chartState = useChartState();

	const {firstDate, lastDate} = useDateTitle();

	const isPreviousPeriodButtonDisabled = useIsPreviousPeriodButtonDisabled();

	const [, addHistoricalWarning] = useHistoricalWarning();

	const addDataSetItems = useAddDataSetItems();

	const changeTimeSpanKey = useChangeTimeSpanKey();

	const nextTimeSpan = useNextTimeSpan();

	const previousTimeSpan = usePreviousTimeSpan();

	const setLoading = useSetLoading();

	const dateFormatters = useMemo(() => dateFormat(languageTag), [
		languageTag,
	]);

	const title = dateFormatters.formatChartTitle([firstDate, lastDate]);

	const isMounted = useIsMounted();

	useEffect(() => {
		let gone = false;

		setLoading();

		const timeSpanComparator =
			chartState.timeSpanKey === LAST_24_HOURS
				? HOUR_IN_MILLISECONDS
				: DAY_IN_MILLISECONDS;

		const keys = [
			'analyticsReportsHistoricalViews',
			'analyticsReportsHistoricalReads',
		];

		if (validAnalyticsConnection) {
			const promises = dataProviders.map((getter) => {
				return getter({
					timeSpanKey: chartState.timeSpanKey,
					timeSpanOffset: chartState.timeSpanOffset,
				});
			});

			allSettled(promises).then((data) => {
				if (gone || !isMounted()) {
					return;
				}

				var dataSetItems = {};

				for (var i = 0; i < data.length; i++) {
					if (data[i].status === 'fulfilled') {
						dataSetItems = {
							...dataSetItems,
							...data[i].value,
						};
					}
					else {
						addHistoricalWarning();
					}
				}

				addDataSetItems({
					dataSetItems,
					keys,
					timeSpanComparator,
				});
			});
		}
		else {
			addDataSetItems({
				keys,
				timeSpanComparator,
			});
		}

		return () => {
			gone = true;
		};
		// eslint-disable-next-line react-hooks/exhaustive-deps
	}, [chartState.timeSpanKey, chartState.timeSpanOffset]);

	const {dataSet} = chartState;
	const {histogram, keyList} = dataSet;

	const referenceDotPosition = useMemo(() => {
		const publishDateISOString = new Date(publishDate).toISOString();

		return chartState.timeSpanKey === LAST_24_HOURS
			? publishDateISOString.split(':')[0].concat(':00:00')
			: publishDateISOString.split('T')[0].concat('T00:00:00');
	}, [chartState.timeSpanKey, publishDate]);

	const handleTimeSpanChange = (event) => {
		const {value} = event.target;

		changeTimeSpanKey({key: value});
	};

	const legendFormatter =
		dataSet &&
		legendFormatterGenerator(
			dataSet.totals,
			languageTag,
			publishedToday,
			validAnalyticsConnection
		);

	const xAxisFormatter =
		chartState.timeSpanKey === LAST_24_HOURS
			? dateFormatters.formatNumericHour
			: dateFormatters.formatNumericDay;

	const lineChartWrapperClasses = className('line-chart-wrapper', {
		'line-chart-wrapper--loading': chartState.loading,
	});

	return (
		<>
			{timeSpanOptions.length && (
				<div className="c-mb-3 c-mt-4">
					<TimeSpanSelector
						disabledNextTimeSpan={chartState.timeSpanOffset === 0}
						disabledPreviousPeriodButton={
							isPreviousPeriodButtonDisabled
						}
						onNextTimeSpanClick={nextTimeSpan}
						onPreviousTimeSpanClick={previousTimeSpan}
						onTimeSpanChange={handleTimeSpanChange}
						timeSpanKey={chartState.timeSpanKey}
						timeSpanOptions={timeSpanOptions}
					/>
				</div>
			)}

			{dataSet ? (
				<div className={lineChartWrapperClasses}>
					{chartState.loading && (
						<ClayLoadingIndicator
							className="chart-loading-indicator"
							small
						/>
					)}

					{title && <h5 className="mb-3">{title}</h5>}

					<div className="line-chart">
						<LineChart
							data={histogram}
							height={CHART_SIZES.height}
							width={CHART_SIZES.width}
						>
							<Legend
								formatter={legendFormatter}
								iconSize={0}
								layout="vertical"
								verticalAlign="top"
								wrapperStyle={{
									left: 0,
									top: 0,
								}}
							/>

							<CartesianGrid
								stroke={CHART_COLORS.cartesianGrid}
								strokeDasharray="0 0"
								vertical={true}
								verticalPoints={[
									CHART_SIZES.width - CHART_SIZES.dotRadius,
								]}
							/>

							<XAxis
								axisLine={{
									stroke: CHART_COLORS.cartesianGrid,
								}}
								dataKey="label"
								domain={
									!validAnalyticsConnection ||
									histogram.length === 0
										? [
												new Date(
													chartState.timeRange.startDate
												).getDate(),
												new Date(
													chartState.timeRange.endDate
												).getDate(),
										  ]
										: []
								}
								interval="preserveStartEnd"
								tickCount={7}
								tickFormatter={(value) => {
									return validAnalyticsConnection &&
										histogram.length !== 0
										? xAxisFormatter(value)
										: value;
								}}
								tickLine={false}
								type={
									validAnalyticsConnection &&
									histogram.length !== 0
										? 'category'
										: 'number'
								}
							/>

							{!validAnalyticsConnection ||
							publishedToday ||
							histogram.length === 0 ? (
								<YAxis
									axisLine={{
										stroke: CHART_COLORS.cartesianGrid,
									}}
									tickLine={false}
									ticks={[0, 50, 100]}
									width={CHART_SIZES.yAxisWidth}
								/>
							) : (
								<YAxis
									allowDecimals={false}
									axisLine={{
										stroke: CHART_COLORS.cartesianGrid,
									}}
									minTickGap={3}
									tickFormatter={thousandsToKilosFormater}
									tickLine={false}
									width={CHART_SIZES.yAxisWidth}
								/>
							)}

							{validAnalyticsConnection && !publishedToday && (
								<Tooltip
									animationDuration={0}
									content={
										<CustomTooltip
											publishDateFill={
												CHART_COLORS.publishDate
											}
											showPublishedDateLabel={
												isPreviousPeriodButtonDisabled
											}
										/>
									}
									cursor={
										validAnalyticsConnection &&
										histogram.length !== 0 &&
										!publishedToday
									}
									formatter={(value, name) => {
										return [
											numberFormat(languageTag, value),
											keyToTranslatedLabelValue(name),
											keyToIconType(name),
										];
									}}
									labelFormatter={
										dateFormatters.formatLongDate
									}
									separator={': '}
								/>
							)}

							{keyList.map((keyName) => {
								const color = keyToHexColor(keyName);
								const shape = keyToIconType(keyName);

								return (
									<Line
										activeDot={
											<CustomActiveDot shape={shape} />
										}
										dataKey={keyName}
										dot={<CustomDot shape={shape} />}
										fill={color}
										isAnimationActive={false}
										key={keyName}
										stroke={color}
										strokeWidth={CHART_SIZES.lineWidth}
										type="monotone"
									/>
								);
							})}

							{validAnalyticsConnection && !publishedToday && (
								<ReferenceDot
									isFront={true}
									r={4}
									stroke={CHART_COLORS.publishDate}
									strokeWidth={CHART_SIZES.lineWidth}
									x={referenceDotPosition}
									y={0}
								/>
							)}
						</LineChart>
					</div>
				</div>
			) : null}
		</>
	);
}

function allSettled(promises) {
	return Promise.all(
		promises.map((promise) => {
			return promise
				.then((value) => {
					return {status: 'fulfilled', value};
				})
				.catch((reason) => {
					return {reason, status: 'rejected'};
				});
		})
	);
}

Chart.propTypes = {
	dataProviders: PropTypes.arrayOf(PropTypes.func).isRequired,
	languageTag: PropTypes.string.isRequired,
	publishDate: PropTypes.string.isRequired,
	timeSpanOptions: PropTypes.arrayOf(
		PropTypes.shape({
			key: PropTypes.string.isRequired,
			label: PropTypes.string.isRequired,
		})
	).isRequired,
};
