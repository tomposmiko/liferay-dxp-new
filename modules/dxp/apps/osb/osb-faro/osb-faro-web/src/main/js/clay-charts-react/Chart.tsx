import BillboardChart from 'react-billboardjs';
import GeoMap from './GeoMap';
import Predictive from './Predictive';
import React from 'react';

import {config} from 'clay-charts-shared';

const {
	DEFAULT_COLORS,
	DEFAULT_GRID_OBJECT,
	DEFAULT_LINE_CLASSES,
	DEFAULT_POINT_PATTERNS
} = config;

/**
 * Chart component.
 * @param {Object} props
 * @return {ReactElement}
 */

interface IPropsCharts {
	color?: string;
	data: any;
	grid?: boolean;
	line?: boolean;
	point?: boolean;
}

type RefType = {
	ref: string;
};

export default React.forwardRef<RefType, IPropsCharts>((props, ref) => {
	const {color, data, grid, line, point, ...otherProps} = props;

	let ChartComponent;

	switch (data.type) {
		case 'geo-map':
			ChartComponent = GeoMap;
			break;
		case 'predictive':
			ChartComponent = Predictive;
			break;
		default:
			ChartComponent = BillboardChart;
	}

	return (
		<ChartComponent
			{...otherProps}
			color={Object.assign({pattern: DEFAULT_COLORS}, color)}
			data={data}
			grid={Object.assign(DEFAULT_GRID_OBJECT, grid)}
			line={Object.assign({classes: DEFAULT_LINE_CLASSES}, line)}
			point={Object.assign(
				{
					pattern: DEFAULT_POINT_PATTERNS
				},
				point
			)}
			ref={ref}
		/>
	);
});
