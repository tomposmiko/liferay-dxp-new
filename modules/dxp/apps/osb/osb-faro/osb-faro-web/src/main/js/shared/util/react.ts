import {isEqual} from 'lodash';

export const getDisplayName = WrappedComponent =>
	WrappedComponent.displayName || WrappedComponent.name || 'Component';

type HasChanges = <T>(
	prev: T | object,
	next: T | object,
	...keys: string[]
) => boolean;

/**
 * Compare previous state or props object by provided keys to detect changes.
 */
export const hasChanges: HasChanges = (prev = {}, next = {}, ...keys) => {
	for (const key of keys) {
		if (key in next) {
			const newVal = next[key];

			const prevVal = prev[key];

			if (!isEqual(newVal, prevVal)) {
				return true;
			}
		}
	}

	return false;
};
