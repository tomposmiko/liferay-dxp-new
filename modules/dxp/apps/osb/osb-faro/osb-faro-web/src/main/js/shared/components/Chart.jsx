import autobind from 'autobind-decorator';
import ClayChart from 'clay-charts-react';
import getCN from 'classnames';
import NoResultsDisplay from 'shared/components/NoResultsDisplay';
import omitDefinedProps from 'shared/util/omitDefinedProps';
import React from 'react';
import Spinner from 'shared/components/Spinner';
import {defer, get, merge} from 'lodash';
import {hasChanges} from 'shared/util/react';
import {PropTypes} from 'prop-types';

const DEFAULT_HEIGHT = 328;

export const AREA = 'area';
export const BAR_CHART = 'bar';
export const COMBINED_CHART = 'combined';
export const DONUT_CHART = 'donut';
export const GAUGE_CHART = 'gauge';
export const LINE_CHART = 'line';
export const SPLINE_CHART = 'spline';

export const CHART_COLORS = [
	'#4B9BFF',
	'#FFB46E',
	'#FF5F5F',
	'#50D2A0',
	'#FF73C3',
	'#9CE269',
	'#B077FF',
	'#FFD76E',
	'#5FC8FF'
];

export const CHART_COLOR_NAMES = {
	greyjoy: '#000000',
	lannister: '#FF5F5F',
	martell: '#50D2A0',
	martellD2: '#31BE88',
	martellD4: '#26966B',
	martellL1: '#64D7AB',
	martellL2: '#79DCB6',
	martellL4: '#A1E7CC',
	mormont: '#FFB46E',
	mormontD2: '#FF9A3B',
	mormontL2: '#FFCEA1',
	mormontL4: '#FFE9D4',
	stark: '#4B9BFF',
	starkD2: '#187FFF',
	starkL2: '#7EB7FF',
	starkL4: '#B1D4FF'
};

const CHART_TYPES = [
	AREA,
	BAR_CHART,
	COMBINED_CHART,
	DONUT_CHART,
	GAUGE_CHART,
	LINE_CHART,
	SPLINE_CHART
];

/**
 * Convert previous api to current api.
 * @param {Array} columnData
 */
function convertData(columnData, chartType = COMBINED_CHART) {
	return columnData.reduce(
		(acc, {axis, color, data, id, type}) => {
			acc.columns = acc.columns.concat([[id, ...data]]);

			if (chartType === COMBINED_CHART) {
				if (!acc.types) {
					acc.types = {};
				}

				acc.types = {...acc.types, [id]: type};
			} else {
				acc.type = chartType;
			}

			if (axis) {
				acc.axes = {
					...acc.axes,
					[id]: axis
				};
			}

			if (color) {
				if (!acc.colors) {
					acc.colors = {};
				}

				acc.colors = {...acc.colors, [id]: color};
			}

			return acc;
		},
		{axes: {}, columns: []}
	);
}

export default class Chart extends React.Component {
	static defaultProps = {
		alwaysShowSelectedTooltip: false,
		axisRotated: false,
		height: DEFAULT_HEIGHT,
		legend: {},
		line: {},
		otherData: {},
		point: {},
		unloadBeforeLoad: true
	};

	static propTypes = {
		alwaysShowSelectedTooltip: PropTypes.bool,
		axisRotated: PropTypes.bool,
		axisX: PropTypes.object,
		axisY: PropTypes.object,
		axisY2: PropTypes.object,
		chartType: PropTypes.oneOf(CHART_TYPES),
		className: PropTypes.string,
		data: PropTypes.array.isRequired,
		dataId: PropTypes.string.isRequired,
		generateChartOnLoad: PropTypes.bool,
		grid: PropTypes.object,
		height: PropTypes.number,
		id: PropTypes.string.isRequired,
		initialSelectedIndexes: PropTypes.array,
		legend: PropTypes.object,
		line: PropTypes.object,
		loading: PropTypes.bool,
		noResultsProps: PropTypes.object,
		onPointSelect: PropTypes.func,
		otherData: PropTypes.object,
		point: PropTypes.object,
		unloadBeforeLoad: PropTypes.bool,
		x: PropTypes.string,
		y2Label: PropTypes.string,
		yLabel: PropTypes.string
	};

	constructor(props) {
		super(props);

		this._chartRef = React.createRef();
	}

	componentDidMount() {
		window.addEventListener('toggleSidebar', this.flushChart);
	}

	shouldComponentUpdate(nextProps) {
		return hasChanges(this.props, nextProps, 'data', 'loading');
	}

	componentDidUpdate(prevProps) {
		const {generateChartOnLoad, initialSelectedIndexes} = this.props;

		if (hasChanges(prevProps, this.props, 'data')) {
			this.updateSelectedIndexes(initialSelectedIndexes);

			if (generateChartOnLoad) {
				this.generateChart();
			}
		}
	}

	componentWillUnmount() {
		window.removeEventListener('toggleSidebar', this.flushChart);

		if (this._deferredSelect) {
			clearTimeout(this._deferredSelect);
		}
	}

	@autobind
	flushChart() {
		const chart = this.getChartRef();

		if (chart) {
			chart.flush(true);
		}

		this.handleShowTooltip();
	}

	generateChart() {
		const billboardReact = get(this._chartRef, ['current']);

		if (billboardReact) {
			billboardReact.generateChart();
		}
	}

	@autobind
	handleChartSelect(point) {
		const {onPointSelect} = this.props;

		onPointSelect && onPointSelect(point);
	}

	getChartRef() {
		return get(this._chartRef, ['current', 'chart']);
	}

	/**
	 * Only show tooltip when a single item is selected
	 */
	@autobind
	handleShowTooltip() {
		if (this.props.alwaysShowSelectedTooltip) {
			const chart = this.getChartRef();

			const selectedPoints = chart.selected();

			const multiSelectEnabled =
				chart.internal.config.data_selection_multiple;

			if (selectedPoints.length && !multiSelectEnabled) {
				chart.tooltip.show({
					index: selectedPoints[0].index
				});
			}
		}
	}

	@autobind
	updateSelectedIndexes(indexes = []) {
		const {dataId, id} = this.props;

		// Using `defer` so `chart` will be defined.
		this._deferredSelect = defer(() => {
			const chart = this.getChartRef();

			if (chart) {
				chart.select(`${id}${dataId}`, indexes, true);
			}
		});
	}

	/**
	 * Public method for manually selecting a point. This is often used for the
	 * first render, which is why the `defer` is necessary, otherwise the
	 * `chart` is undefined.
	 * @param {Array} indexes - The indexes you want to select
	 */
	select(indexes) {
		const {dataId, id} = this.props;

		this._deferredSelect = defer(() => {
			const chart = this.getChartRef();

			if (chart) {
				chart.select(`${id}${dataId}`, indexes);
			}

			this.handleShowTooltip();
		});
	}

	/**
	 * Public method for manually unselecting a point.
	 * @param {Array} indexes - The indexes you want to unselect
	 */
	unselect(indexes) {
		const {alwaysShowSelectedTooltip, dataId, id} = this.props;

		const chart = this.getChartRef();

		if (chart) {
			chart.unselect(`${id}${dataId}`, indexes);

			if (alwaysShowSelectedTooltip) {
				chart.tooltip.hide();
			}
		}
	}

	renderChart() {
		const {
			axisRotated,
			axisX,
			axisY,
			axisY2,
			chartType,
			data,
			grid,
			height,
			legend,
			line,
			loading,
			noResultsProps,
			onPointSelect,
			otherData,
			point,
			unloadBeforeLoad,
			x,
			...otherProps
		} = this.props;

		if (data.length) {
			return (
				<>
					{/* eslint-disable */}
					<ClayChart
						{...omitDefinedProps(otherProps, Chart.propTypes)}
						axis={{
							rotated: axisRotated,
							x: merge(
								{tick: {outer: false, show: false}},
								axisX
							),
							y: merge(
								{tick: {outer: false, show: false}},
								axisY
							),
							y2: merge(
								{tick: {outer: false, show: false}},
								axisY2
							)
						}}
						data={{
							...convertData(data, chartType),
							onclick: this.handleChartSelect,
							selection: onPointSelect
								? {
										enabled: true,
										grouped: true,
										multiple: false
								  }
								: {},
							x,
							...otherData
						}}
						grid={merge({x: {show: false}, y: {show: true}}, grid)}
						legend={merge({show: false}, legend)}
						line={merge({classes: null}, line)}
						onout={this.handleShowTooltip}
						point={merge(
							{
								focus: {expand: {enabled: true, r: 5}},
								pattern: ['circle'],
								r: 0,
								select: {r: 5}
							},
							point
						)}
						ref={this._chartRef}
						selection={
							onPointSelect
								? {
										enabled: true,
										grouped: true,
										multiple: false
								  }
								: {}
						}
						unloadBeforeLoad={unloadBeforeLoad}
						size={{height}}
					/>
					{/* eslint-enable */}
				</>
			);
		} else if (!loading) {
			return <NoResultsDisplay {...noResultsProps} />;
		}
	}

	render() {
		const {className, data, loading, onPointSelect, y2Label, yLabel} =
			this.props;

		const classes = getCN('chart-root', className, {
			selectable: onPointSelect
		});

		return (
			<div className={classes}>
				{(yLabel || y2Label) && (
					<div className='chart-labels'>
						<span>{yLabel}</span>

						<span>{y2Label}</span>
					</div>
				)}

				{this.renderChart()}

				{loading && (
					<Spinner overlay={!!data.length} spacer={!data.length} />
				)}
			</div>
		);
	}
}
