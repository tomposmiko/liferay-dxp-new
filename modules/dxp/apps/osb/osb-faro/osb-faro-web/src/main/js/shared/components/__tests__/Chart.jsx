import Chart from '../Chart';
import React from 'react';
import {cleanup, render} from '@testing-library/react';

jest.unmock('react-dom');

describe('Chart', () => {
	afterEach(cleanup);

	it('should render', () => {
		const {container} = render(
			<Chart data={[{data: [], id: 'foo'}]} dataId='bar' id='foo' />
		);
		expect(container).toMatchSnapshot();
	});

	it('should render w/ NoResultsDisplay', () => {
		const {container} = render(<Chart data={[]} dataId='bar' id='foo' />);

		expect(container.querySelector('.no-results-root')).toBeTruthy();
	});

	it('should render w/ loading spinner', () => {
		const {container} = render(
			<Chart data={[]} dataId='bar' id='foo' loading />
		);

		expect(container.querySelector('.spinner-root')).toBeTruthy();
	});
});
