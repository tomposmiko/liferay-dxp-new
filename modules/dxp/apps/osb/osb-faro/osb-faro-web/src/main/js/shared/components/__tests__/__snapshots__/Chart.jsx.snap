// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chart should render 1`] = `
<div>
  <div
    class="chart-root"
  >
    <div />
  </div>
</div>
`;
